op-batcher/batcher/espresso.go

```diff
@@ -0,0 +1,163 @@
+	"fmt"
+	"time"
+
+	espressoVerification "github.com/EspressoSystems/espresso-sequencer-go/verification"
+
+	espressoCommon "github.com/EspressoSystems/espresso-sequencer-go/types"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// TODO: Pull out to be re-used in op-node for derivation from Espresso
+type Transaction struct {
+	// Namespace of transaction to be published
+	Namespace uint64
+	// TEE attestation to be verified by op-node
+	TeeAttn []byte
+	// Frames serialized as they would be for posting to L1 as calldata
+	CallData []byte
+}
+
+func (t Transaction) toEspresso() espressoCommon.Transaction {
+	payload := append(t.TeeAttn, t.CallData...)
+	return espressoCommon.Transaction{
+		Namespace: t.Namespace,
+		Payload:   payload,
+	}
+}
+
+func (l *BatchSubmitter) waitForFinality(height uint64, rawHeader json.RawMessage, header *espressoCommon.HeaderImpl) error {
+	timer := time.NewTimer(2 * time.Minute)
+	defer timer.Stop()
+
+	ticker := time.NewTicker(100 * time.Millisecond)
+	defer ticker.Stop()
+
+	var snapshot espressoCommon.BlockMerkleSnapshot
+
+Loop:
+	for {
+		select {
+		case <-ticker.C:
+			res, err := l.EspressoLightClient.FetchMerkleRoot(height, nil)
+			if err == nil {
+				snapshot = res
+				break Loop
+			}
+		case <-timer.C:
+			return fmt.Errorf("failed to fetch merkle root")
+		}
+	}
+
+	if snapshot.Height <= height {
+		return fmt.Errorf("snapshot height is less than or equal to the requested height")
+	}
+
+	nextHeader, err := l.Espresso.FetchHeaderByHeight(l.shutdownCtx, snapshot.Height)
+	if err != nil {
+		return fmt.Errorf("error fetching the snapshot header (height: %d): %w", snapshot.Height, err)
+	}
+
+	proof, err := l.Espresso.FetchBlockMerkleProof(l.shutdownCtx, snapshot.Height, height)
+	if err != nil {
+		return fmt.Errorf("error fetching merkle proof")
+	}
+
+	blockMerkleTreeRoot := nextHeader.Header.GetBlockMerkleTreeRoot()
+
+	log.Info("Verifying merkle proof", "height", height)
+	ok := espressoVerification.VerifyMerkleProof(proof.Proof, rawHeader, *blockMerkleTreeRoot, snapshot.Root)
+	if !ok {
+		return fmt.Errorf("error validating merkle proof (height: %d, snapshot height: %d)", height, snapshot.Height)
+	}
+
+	// Verify the namespace proof
+	log.Info("Verifying namespace proof", "height", height)
+	resp, err := l.Espresso.FetchTransactionsInBlock(l.shutdownCtx, height, 42)
+	if err != nil {
+		return fmt.Errorf("failed to fetch the transactions in block")
+	}
+
+	namespaceOk := espressoVerification.VerifyNamespace(
+		42,
+		resp.Proof,
+		*header.Header.GetPayloadCommitment(),
+		*header.Header.GetNsTable(),
+		resp.Transactions,
+		resp.VidCommon,
+	)
+
+	if !namespaceOk {
+		return fmt.Errorf("error validating namespace proof (height: %d)", height)
+	}
+
+	return nil
+}
+
+func (l *BatchSubmitter) submitToEspresso(txdata txData) (*EspressoCommitment, error) {
+	transaction := Transaction{
+		Namespace: 42,
+		TeeAttn:   []byte{1, 2, 3, 4},
+		CallData:  txdata.CallData(),
+	}.toEspresso()
```

Comment:
Nit: It's fine as a proof of concept, but could we define this `transaction` somewhere at the beginning of this file and make it a const like `TestTransaction`? Just so we know we are going to replace it.